packages/aws-cdk-lib/custom-resources/lib/provider-framework/waiter-state-machine.ts

```diff
@@ -75,23 +125,77 @@ export class WaiterStateMachine extends Construct {
       },
     });
 
-    const resource = new CfnResource(this, 'Resource', {
-      type: 'AWS::StepFunctions::StateMachine',
-      properties: {
-        DefinitionString: definition,
-        RoleArn: role.roleArn,
-      },
+    this.isCompleteHandler = props.isCompleteHandler;
+    const resource = new CfnStateMachine(this, 'Resource', {
+      definitionString: definition,
+      roleArn: role.roleArn,
+      loggingConfiguration: this.renderLoggingConfiguration(role, props.logOptions, props.disableLogging),
     });
     resource.node.addDependency(role);
 
     this.stateMachineArn = resource.ref;
   }
 
+  /**
+   * Grant the given identity permissions on StartExecution of the state machine.
+   */
   public grantStartExecution(identity: IGrantable) {
     return Grant.addToPrincipal({
       grantee: identity,
       actions: ['states:StartExecution'],
       resourceArns: [this.stateMachineArn],
     });
   }
+
+  private renderLoggingConfiguration(
+    role: Role,
+    logOptions?: LogOptions,
+    disableLogging?: boolean,
+  ): CfnStateMachine.LoggingConfigurationProperty | undefined {
+    if (disableLogging) return undefined;
+
+    // You need to specify `*` in the Resource field because CloudWatch API actions, such as
+    // CreateLogDelivery and DescribeLogGroups, don't support Resource types defined by Amazon
+    // CloudWatch Logs.
+    // https://docs.aws.amazon.com/step-functions/latest/dg/cw-logs.html#cloudwatch-iam-policy
+    role.addToPrincipalPolicy(new PolicyStatement({
+      actions: [
+        'logs:CreateLogDelivery',
+        'logs:CreateLogStream',
+        'logs:GetLogDelivery',
+        'logs:UpdateLogDelivery',
+        'logs:DeleteLogDelivery',
+        'logs:ListLogDeliveries',
+        'logs:PutLogEvents',
+        'logs:PutResourcePolicy',
+        'logs:DescribeResourcePolicies',
+        'logs:DescribeLogGroups',
+      ],
+      resources: ['*'],
+    }));
+
+    const logGroup = logOptions?.destination ?? new LogGroup(this, 'LogGroup', {
+      // Log group name should start with `/aws/vendedlogs/` to not exceed Cloudwatch Logs Resource Policy
+      // size limit.
+      // https://docs.aws.amazon.com/step-functions/latest/dg/bp-cwl.html
+      //
+      // By using the auto-generated name of the Lambda created in the `Provider` that calls this
+      // `WaiterStateMachine` construct, even if the `Provider` (or its parent) is deleted and then
+      // created again, the log group name will not duplicate previously created one with removal
+      // policy `RETAIN`. This is because that the Lambda will be re-created again with auto-generated name.
+      // The `node.addr` is also used to prevent duplicate names no matter how many times this construct
+      // is created in the stack. It will not duplicate if called on other stacks.
+      logGroupName: `/aws/vendedlogs/states/waiter-state-machine-${this.isCompleteHandler.functionName}-${this.node.addr}`,
+    });
```

Comment:
This code fixes [this issue](https://github.com/aws/aws-cdk/issues/28577).

The other codes are basically the same as [this PR](https://github.com/aws/aws-cdk/pull/27310) which has been reverted.